compiler/rustc_const_eval/src/interpret/machine.rs

```diff
@@ -285,11 +289,12 @@ pub trait Machine<'mir, 'tcx>: Sized {
         addr: u64,
     ) -> Pointer<Option<Self::PointerTag>>;
 
-    /// Convert a pointer with provenance into an allocation-offset pair.
+    /// Convert a pointer with provenance into an allocation-offset pair
+    /// and extra provenance info.
     fn ptr_get_alloc(
         ecx: &InterpCx<'mir, 'tcx, Self>,
         ptr: Pointer<Self::PointerTag>,
-    ) -> (AllocId, Size);
+    ) -> (AllocId, Size, Self::TagExtra);
```

Comment:
Technically, this is currently still redundant -- we could make this `ptr_get_offset` and have it return just a `Size`, and get all the other data directly from the `PointerTag`.

But in preparation for @carbotaniuman's PR, I think we should leave this the way it is. After that PR, `Provenance::get_alloc_id` should be called only for diagnostics and when `Provenance::OFFSET_IS_ADDR == false`, and `ptr_get_alloc` should be used throughout the machine to extract information from provenance.